### PR TITLE
Extract append-only optimization from Engine

### PR DESCRIPTION
**Pull request by nik9000**
This extracts the logic for the "append only" optimization from `Engine`
into a pluggable behavior class so that we can override it in TSDB.

### CODE REVIEW

**Comment by jpountz** [comment]
Out of curiosity, what would the implementation look like for TSDB? Would it track the max timestamp for each timeseries or something like that?

**Comment by nik9000** [comment]
> Out of curiosity, what would the implementation look like for TSDB? Would it track the max timestamp for each timeseries or something like that?

I was thinking of keeping some number of max timestamps, yeah. Like grabbing the low nibble from the hashed tsid and storing 64 timestamps. Or something like that. Then I got distracted by that `_id` query optimization. Then I actually merged the id generation and had a bunch of follow up to run through.

**Comment by jpountz** [comment]
I just had a quick chat with @henningandersen about this PR and I believe that we have two main options at a high level. The first one is the one that you are suggesting, that consists of extracting the timestamp from the ID to optimize writes in case this timestamp is greater than all other timestamps that have been seen before. And the second one consists of generating IDs so that Elasticsearch and Lucene could do it mostly automatically by having an ID that concatenates the timestamp in BE order and then the TSID.

The trade-off I'm seeing is that your approach require more maintenance and is specific to timeseries data streams, but it is likely more space-efficient thanks to better sharing of prefixes of IDs, and more CPU efficient in case data comes in timestamp order on a per-timeseries basis but not globally (which is not unlikely?).

I'm curious if you have any data about how much larger the index of the _id field would be if we generated the ID by putting the timestamp first instead of last?

**Comment by jpountz** [comment]
At the moment my gut feeling is that we'd need both too. I'm sort of hoping that data about disk usage would confirm that it's not ok so that I feel more confident about moving forward in the direction you suggest.

**Comment by nik9000** [comment]
> At the moment my gut feeling is that we'd need both too.

:+1: I'm glad our stomachs agree. I'll try and get data. Though I might be getting distracted in the short term.

File: server/src/main/java/org/elasticsearch/index/engine/MayHaveBeenIndexedBefore.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    /**
+     * Returns {@code true} if the indexing operation may have already be
+     * processed by the engine. Note that it is OK to rarely return true even
+     * if this is not the case. However a {@code false} return value must
+     * always be correct.
+     * <p>
+     * This relies on state internal to the implementation and may modify
+     * that state.
+     */
+    boolean mayHaveBeenIndexedBefore(Index index);
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/index/engine/MayHaveBeenIndexedBefore.java:48]: Can we collapse those into one method? I see a conflict to the assertion in `indexIntoLucene`, but I think I would prefer to make that an explicit `assertXYZ` method instead then.

**Review by henningandersen — COMMENTED** [review]
A few comments from an initial read.

> I was thinking of keeping some number of max timestamps, yeah. Like grabbing the low nibble from the hashed tsid and storing 64 timestamps.

IIRC, we could have millions of tsids. The way we make this safe todahy in failover cases is to ensure the replica knows the max unsafe timestamp. For tsdb, it would need to know the max timestamp or we would have to bootstrap that from scratch on failover. Is the latter your thought here?

I wonder if that necessarily goes well, since then on failover, the first request for every tsid will have an extended duration. It might need the normal "does this id exist" check and/or also have to search the tsid to find the largest timestamp.

If we imagine a cluster max'ed out (more or less) indexing with the optimization, it might fall over when a node dies? Or at least it might buffer up loads of data, reject some and require retries from clients. This could even be the case for some relocations maybe, in particular if they have just one data stream with many tsids.

**Comment by martijnvg** [comment]
The index operation results for the tsdb track when running with this change:

```
|                                                Min Throughput |                   index | 35402.8         | 36538.3         | 1135.55    | docs/s |   +3.21% |
|                                               Mean Throughput |                   index | 37938.5         | 40071.2         | 2132.69    | docs/s |   +5.62% |
|                                             Median Throughput |                   index | 37776           | 40121.3         | 2345.31    | docs/s |   +6.21% |
|                                                Max Throughput |                   index | 42319           | 44990.3         | 2671.25    | docs/s |   +6.31% |
|                                       50th percentile latency |                   index |   968.057       |   908.885       |  -59.1719  |     ms |   -6.11% |
|                                       90th percentile latency |                   index |  1382.28        |  1319.74        |  -62.5386  |     ms |   -4.52% |
|                                       99th percentile latency |                   index |  3577.88        |  3486.66        |  -91.2231  |     ms |   -2.55% |
|                                     99.9th percentile latency |                   index |  6542.95        |  6145.17        | -397.778   |     ms |   -6.08% |
|                                    99.99th percentile latency |                   index |  8484.75        |  7568.61        | -916.14    |     ms |  -10.80% |
|                                      100th percentile latency |                   index |  9014.89        |  8420.02        | -594.871   |     ms |   -6.60% |
|                                  50th percentile service time |                   index |   968.057       |   908.885       |  -59.1719  |     ms |   -6.11% |
|                                  90th percentile service time |                   index |  1382.28        |  1319.74        |  -62.5386  |     ms |   -4.52% |
|                                  99th percentile service time |                   index |  3577.88        |  3486.66        |  -91.2231  |     ms |   -2.55% |
|                                99.9th percentile service time |                   index |  6542.95        |  6145.17        | -397.778   |     ms |   -6.08% |
|                               99.99th percentile service time |                   index |  8484.75        |  7568.61        | -916.14    |     ms |  -10.80% |
|                                 100th percentile service time |                   index |  9014.89        |  8420.02        | -594.871   |     ms |   -6.60% |
```

Just attaching this here for keeping record of this. I ran with the benchmark defaults (indexing into a tsdb index, 8 clients concurrently indexing).

**Comment by nik9000** [comment]
This one's so stale it won't go in. We might be able to reuse parts of it one day, but no need to keep it open.

**Pull request closed without merging**